### PR TITLE
Exchange guard cells with the electrostatic solver

### PR DESCRIPTION
**Pull request by RemiLehe**
It seems that we forgot to exchange guard cells after the Poisson solver, in the case of the electrostatic solver.

This was detected by looking at energy conservation in a periodic box with a thermal plasma: 
[input_2d.txt](https://github.com/ECP-WarpX/WarpX/files/14363054/input_2d.txt)
Screenshot 2024-02-21 at 9 36 43 AM

I added a smaller version of the above test as an automated test. The test checks that the energy is within the dashed back line:

Screenshot 2024-02-22 at 2 48 02 PM

Many thanks to @oshapoval and @jlvay for helping in debugging this.

### CODE REVIEW

File: Examples/Tests/energy_conserving_thermal_plasma/analysis.py [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+# Here, we use energy-converving gather and an electrostatic solver. The energy
+# is not expected to be exactly conserved, but it is expected to be better conserved
+# than other gathering scheme. This tests checks that the energy does not increase by
+# more than 0.3% over the duration of the simulatoin.
+
+import os
+import sys
+
+import numpy as np
+import yt
```
Comment (RemiLehe) [inline Examples/Tests/energy_conserving_thermal_plasma/analysis.py]:
```suggestion
```

File: Regression/WarpX-tests.ini [hunk 2]
```diff
@@ -4627,3 +4627,19 @@ compileTest = 0
 doVis = 0
 compareParticles = 1
 analysisRoutine = Examples/Tests/Implicit/analysis_1d.py
+
+[EnergyConservingThermalPlasma]
+buildDir = .
+inputFile = Examples/Tests/Implicit/inputs_2d_electrostatic
```
Comment (RemiLehe) [inline Regression/WarpX-tests.ini]:
```suggestion
inputFile = Examples/Tests/energy_conserving_thermal_plasma/inputs_2d_electrostatic
```

File: Examples/Tests/energy_conserving_thermal_plasma/inputs_2d_electrostatic [hunk 3]
```diff
@@ -0,0 +1,90 @@
...
+warpx.do_electrostatic = labframe
+algo.field_gathering = energy-conserving
+algo.particle_shape = 2
+warpx.use_filter = 0
+
+#################################
+############ CONSTANTS #############
+#################################
+
+my_constants.n0 = 1.e30          # plasma densirty, m^-3
```
Comment (lucafedeli88) [inline Examples/Tests/energy_conserving_thermal_plasma/inputs_2d_electrostatic]:
```suggestion
my_constants.n0 = 1.e30          # plasma density, m^-3
```
just a typo

File: Examples/Tests/energy_conserving_thermal_plasma/inputs_2d_electrostatic [hunk 4]
```diff
@@ -0,0 +1,90 @@
...
+boundary.field_lo = periodic periodic
+boundary.field_hi = periodic periodic
+boundary.particle_lo = periodic periodic
+boundary.particle_hi = periodic periodic
+
+#################################
+####### Boundary condition ######
+#################################
+boundary.field_lo = periodic periodic
+boundary.field_hi = periodic periodic
```
Comment (lucafedeli88) [inline Examples/Tests/energy_conserving_thermal_plasma/inputs_2d_electrostatic]:
```suggestion

#################################
###### BOUNDARY CONDITIONS ######
#################################
boundary.field_lo = periodic periodic
boundary.field_hi = periodic periodic
boundary.particle_lo = periodic periodic
boundary.particle_hi = periodic periodic
```

The same conditions are repeated twice.

File: Examples/Tests/energy_conserving_thermal_plasma/inputs_2d_electrostatic [hunk 5]
```diff
@@ -0,0 +1,90 @@
...
+
+#################################
+############ NUMERICS ###########
+#################################
+warpx.serialize_initial_conditions = 1
+warpx.verbose = 1
```
Comment (lucafedeli88) [inline Examples/Tests/energy_conserving_thermal_plasma/inputs_2d_electrostatic]: I think that these parameters could be considered "general parameters" rather then "numerics" (they are not technical parameters of the solvers used in WarpX)

File: Examples/Tests/energy_conserving_thermal_plasma/inputs_2d_electrostatic [hunk 6]
```diff
@@ -0,0 +1,90 @@
...
+#################################
+max_step = 500
+amr.n_cell =  8 8
+amr.max_level = 0
+geometry.dims = 2
+
+warpx.do_electrostatic = labframe
+algo.field_gathering = energy-conserving
+algo.particle_shape = 2
+warpx.use_filter = 0
```
Comment (lucafedeli88) [inline Examples/Tests/energy_conserving_thermal_plasma/inputs_2d_electrostatic:12]: I think that these parameters should be under "numerics"

File: Examples/Tests/energy_conserving_thermal_plasma/inputs_2d_electrostatic [hunk 7]
```diff
@@ -0,0 +1,90 @@
...
+my_constants.de0 = clight/wpe    # skin depth, m
+my_constants.dt = ( 0.2 )/wpe        # time step size, s
+warpx.const_dt = dt
+#################################
+####### GENERAL PARAMETERS ######
+#################################
+
+geometry.dims = 2
+geometry.prob_lo = 0.0 0.0
+geometry.prob_hi = 10.*de0 10.*de0
```
Comment (lucafedeli88) [inline Examples/Tests/energy_conserving_thermal_plasma/inputs_2d_electrostatic:30]: These parameters could be placed in the "general parameters" section above.

**Review by lucafedeli88 — CHANGES_REQUESTED** [review]
Thanks for this PR! I left few minor comments

**Review by dpgrote — APPROVED** [review]
Thanks! This looks good.
